tokio-util/src/sync/cancellation_token.rs

```diff
@@ -241,6 +253,46 @@ impl CancellationToken {
     pub fn drop_guard(self) -> DropGuard {
         DropGuard { inner: Some(self) }
     }
+
+    /// Runs a future to completion and returns its result unless the `CompletionToken` is
+    /// cancelled. In that case the function returns `None` and the future gets dropped.
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is not cancel safe since the input future gets dropped on cancel.
+    pub async fn run_until_cancelled<F>(&self, fut: F) -> Option<F::Output>
+    where
+        F: Future,
+    {
+        RunUntilCancelledFuture {
+            cancellation_token: self,
+            future: fut,
+        }
+        .await
+    }
+}
+
+// ===== impl RunUntilCancelledFuture =====
+
+impl<'a, F: Future> core::fmt::Debug for RunUntilCancelledFuture<'a, F> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("RunUntilCancelledFuture").finish()
+    }
+}
+
+impl<'a, F: Future> Future for RunUntilCancelledFuture<'a, F> {
+    type Output = Option<F::Output>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+        if let Poll::Ready(res) = this.future.as_mut().poll(cx) {
+            Poll::Ready(Some(res))
+        } else if this.cancellation_token.is_cancelled() {
+            Poll::Ready(None)
```

Comment:
You probably want `poll` to get called when the token is cancelled, but that won't happen with this implementation. To achieve that, you should store a `WaitForCancellationFuture` inside `RunUntilCancelledFuture` and poll it instead of using `is_cancelled`.

Reply:
Thanks for that! I changed the implementation of `RunUntilCancelledFuture` accordingly.